modules/serviceregistry/src/main/java/org/opencastproject/serviceregistry/impl/ServiceRegistryJpaImpl.java

```diff
@@ -3419,6 +3437,72 @@ public int compare(ServiceRegistration serviceA, ServiceRegistration serviceB) {
 
   }
 
+  /**
+   * Comparator that will sort service registrations depending on their capacity, which is defined by the number of jobs
+   * the service's host is already running divided by the MaxLoad of the Server. The lower that number, the bigger the capacity.
+   * This Comparator will preferre encoding workers, if none are defined in the configuration file it will act like the LoadComparator.
+   */
+  private static final class LoadComparatorEncoding implements Comparator<ServiceRegistration> {
+
+    private SystemLoad loadByHost = null;
+
+    /**
+     * Creates a new comparator which is using the given map of host names and loads.
+     *
+     * @param loadByHost
+     */
+    LoadComparatorEncoding(SystemLoad loadByHost) {
+      this.loadByHost = loadByHost;
+    }
+
+    @Override
+    public int compare(ServiceRegistration serviceA, ServiceRegistration serviceB) {
+      String hostA = serviceA.getHost();
+      String hostB = serviceB.getHost();
+      NodeLoad nodeA = loadByHost.get(hostA);
+      NodeLoad nodeB = loadByHost.get(hostB);
+
+
+
+      if (encodingWorkers != null) {
+
+        if (isEncodingWorker(hostA, encodingWorkers) && !isEncodingWorker(hostB, encodingWorkers)) {
+          if (nodeA.getLoadFactor() <= 0.4) {
```

Comment:
Where did these magic load factors come from?

Reply:
I thought there should be a threshold for loadFactors, below which a specialized worker (short SW.) always should be preferred against a non SW. This is meant to avoid a non SW with loadFactor 0 getting preferred over a SW that still has a low loadFactor of for example 0.1.

The value 0.4 was a first guess. Doing the math, 0.4 allows a SW with 16 cores to have over 16 encoding jobs (à 0.8 JobLoad) before considering a non SW, which seems way to high. Maybe right now SW only should be preferred in idle situations.


I'm still thinking about how to better meet certain guarantees when dispatching, while also properly using the SW and reasonably spreading the load across the system.

Reply:
If the goal here is to load SWs with encoding jobs prior to spilling over to other nodes then how about

```
if nodeA and nodeB are both encoders
  nodeA load + job load >= nodeA's maxload ? nodeB : nodeA
elseif either node is an encoder and that node's load + job load <= node's maxload
  that node
else
  the other node
```

This prefers an encoder, unless all of them are full in which case it picks another.

Reply:
By my understanding the services accepting already check whether the MaxLoad is exceeded, by inheriting the isReadyToAccept function from the AbstractJobProducer. But I think filling the SW up to the MaxLoad before considering non SW, is a to high value when trying to evenly spread the load across the system. With the MaxLoad defaulting to the number of CPU cores that means a SW with 16 Cores could have 16 encoding jobs before considering non SW servers.